jf_agent/main.py

```diff
@@ -184,101 +184,109 @@ def main():
+                logger.error(
+                    f"Failed to run healthcheck validation due to exception, moving on. Exception: {err}"
                 )
 
-            diagnostics.capture_outdir_size(config.outdir)
-
-        except Exception as err:
-            logger.error(f"Encountered error during agent run! {err}")
-            agent_logging.close_out(logging_config)
-            return False
+            try:
+                if jellyfish_endpoint_info.jf_options.get('validate_num_repos', False):
+                    validate_num_repos(config.git_configs, creds)
+            except Exception as e:
+                logger.warning(f"Could not validate client/org creds, moving on. Got {e}")
 
-        finally:
-            # We need to close this before we send data
-            # Otherwise we'll send a .fuse_hidden file (temp file)
-            diagnostics.close_file()
+            try:
+                diagnostics.capture_agent_version()
+                diagnostics.capture_run_args(
+                    args.mode, args.config_file, config.outdir, args.prev_output_dir
+                )
 
-            # Kills the sys_diag_collector thread.
-            # We need to do this before exiting, otherwise we'll hang forever and never exit until timeout kills it.
-            sys_diag_done_event.set()
-            sys_diag_collector.join()
+                if config.run_mode_is_print_apparently_missing_git_repos:
+                    issues_to_scan = get_issues_to_scan_from_jellyfish(
+                        config,
+                        creds,
+                        args.for_print_missing_repos_issues_updated_within_last_x_months,
+                    )
+                    if issues_to_scan:
+                        print_missing_repos_found_by_jira(config, creds, issues_to_scan)
+                    return True
+
+                if config.run_mode_includes_download:
+                    download_data_status = download_data(
+                        config,
+                        creds,
+                        jellyfish_endpoint_info.jira_info,
+                        jellyfish_endpoint_info.git_instance_info,
+                        jellyfish_endpoint_info.jf_options,
+                    )
+                    success = all(s['status'] == 'success' for s in download_data_status)
+                    write_file(
+                        config.outdir, 'status', config.compress_output_files, download_data_status
+                    )
+
+                diagnostics.capture_outdir_size(config.outdir)
+
+            except Exception as err:
+                logger.error(f"Encountered error during agent run! {err}")
+                agent_logging.close_out(logging_config)
+                return False
+
+    finally:
+        logger.info('Closing Diagnostics file')
+        # We need to close this before we send data
+        # Otherwise we'll send a .fuse_hidden file (temp file)
+        diagnostics.close_file()
```

Comment:
Should `diagnostics.close_file()` come after `sys_diag_done_event.join()`... there seems to be a possible race condition here.

Reply:
Also, are we ok with `diagnostics.send_diagnostic_end_reading` potentially not getting called if we return early?  or, should that be part of the finally?

Reply:
> Also, are we ok with `diagnostics.send_diagnostic_end_reading` potentially not getting called if we return early? or, should that be part of the finally?

This one is kind of a toss up for me. If we end becuase of an error and the `heartbeat` stops with no `stop` signal sent than we should be able to quickly tell that it crashed. I could also see sending the `stop` signal with a status attribute working pretty well too... Truthfully the heartbeats aren't really used anywhere yet, this was a thing we shipped as a proof of concept for a hackathon

> Should diagnostics.close_file() come after sys_diag_done_event.join()... there seems to be a possible race condition here.

Yeah good call. Right now it does this and we get a warning message sometimes saying we can't write to the Diagnostics file when we do our final cache flush to it. That could potentially be lost data, so we should close the file after we shut down the thread 👍

Reply:
Sounds good to me.